### ISSUE TITLE
v130 chrome process error on Windows after a long hanging delay when doing mermaid screenshot

### ISSUE THREAD

**Issue by javajon** [quarto-dev/quarto-cli#11135]
### Bug description

Attempting to create a mermaid graph (or dot) in a qmd file and formatting it to PDF results in InvalidCertificate error.

### Steps to reproduce

Given this small qmd in a blank directory with no extensions:

````markdown
```{mermaid}
flowchart LR
  A[Hard edge] --> B(Round edge)
  B --> C{Decision}
  C --> D[Result one]
  C --> E[Result two]
```
````

This command will show the diagram using RevealJS:

````bash
quarto preview test-graph-to-pdf.qmd
````

However, when rendering to PDF it fails:

````bash
quarto preview test-graph-to-pdf.qmd --to pdf
Unable to add system certificate to certificate store: InvalidCertificate(Other(OtherError(UnsupportedCriticalExtension)))
308203b33082029ba00302010202103e9bb16b607c[...]4095e3
````

(I shortened the certificate string output.)

At this point the CLI is hung and must be terminated with CTRL-C.

Also, if I attempt to reinstall chromium I get the same error:

````bash
$ quarto uninstall chromium
? Are you sure you'd like to remove chromium? (Y/n) » Yes

Uninstalling chromium
[>] Removing Chromium...
Uninstallation successful
````

```bash
$ quarto install chromium
Installing chromium
Unable to add system certificate to certificate store: InvalidCertificate(Other(OtherError(UnsupportedCriticalExtension)))
308203b33082029ba00302010202103e9bb16b607c[...]4095e3
[>] Downloading Chromium 869685
[>] Installing Chromium 869685
Installation successful
```

### Expected behavior

PDF preview

### Actual behavior

Error message and CLI is hung.

### Your environment

- Windows 11
- VSCode 1.94.2
- Version 130.0.6723.59 (Official Build) (64-bit)

### Quarto check output

```bash
Quarto 1.6.28
[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.4.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.46.3: OK
      Typst version 0.11.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.6.28
      Path: C:\Users\jonat\AppData\Local\Programs\Quarto\bin       
      CodePage: 1252

[>] Checking tools....................OK
      TinyTeX: v2024.10
      Chromium: 869685

[>] Checking LaTeX....................OK
      Using: TinyTex
      Path: C:\Users\jonat\AppData\Roaming\TinyTeX\bin\windows\    
      Version: 2024

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.12.6
      Path: C:/Users/jonat/AppData/Local/Programs/Python/Python312/python.exe
      Jupyter: 5.7.1
      Kernels: python3

[>] Checking Jupyter engine render....OK

[>] Checking R installation...........OK
      Version: 4.4.1
      Path: C:/PROGRA~1/R/R-44~1.1
      LibPaths:
        - C:/Users/jonat/AppData/Local/R/win-library/4.4
        - C:/Program Files/R/R-4.4.1/library
      knitr: 1.48
      rmarkdown: 2.28

[>] Checking Knitr engine render......OK
```

**Comment by cscheid** [comment]
@cderv can you repro this?

Unfortunately, I believe this only happens if your IT sets up a root certificate that is not being properly recognized. I'd like for Quarto not to freeze, but we can't do much about it until we can repro this locally ourselves.

**Comment by cderv** [comment]
No I can't repro this.

We do set the CA STORE to be used to be system and mozilla 
https://github.com/quarto-dev/quarto-cli/blob/0183439d37e10b713270ae6e901a0ba5d8ddb0ad/package/launcher/src/main.rs#L77

This is done so that it can use local certificate for private organization network I think. It is a Deno env var : https://docs.deno.com/runtime/reference/env_variables/#special-environment-variables

It will search system first, and so if there is problem with system CA store, then we can't do much I think. 

The hang / Panic is deno. 


Related issues on this: 
- https://github.com/denoland/deno/issues/24137

In 1.46.0, they stopped the panicking to throw an error, then a warning and continuing with hex representation: 
- https://github.com/denoland/deno/commit/8e6b06b89d47c0ce27c2f959be6d8d48ffab5ebf 
- https://github.com/denoland/deno/commit/25bb59d2ce83b03356a86013bb80107d29737b52

But it seems it does not solve completely their issue and caused the hand. But I can't see a new issue opened. 

@cscheid maybe we can offer users with problem a way to set `DENO_TLS_CA_STORE` to another value ? 

I believe this is an issue in Deno having problem with some certificate in the System CA store on windows, and this impact Quarto because we do set `DENO_TLS_CA_STORE=system,mozilla`. Maybe allowing to set `DENO_TLS_CA_STORE=mozilla` only would help ? 🤷‍♂

**Comment by javajon** [comment]
Same error with:

```bash
export DENO_TLS_CA_STORE=mozilla

quarto preview test-graph-to-pdf.qmd --to pdf
Unable to add system certificate to certificate store: InvalidCertificate(Other(OtherError(UnsupportedCriticalExtension)))
308203b33082029ba003020102021[...]e384095e3
```

Yes, in previous versions of Quarto with older Deno it was panicking. It's the same error, now instead of panic and dump info, it dumps the error and hangs for a few minutes then dumps this strack trace:

```bash
ERROR: Couldn't find open server.
Chrome process error:

Stack trace:
Chrome process error:
    at criClient (file:///C:/Users/jonat/AppData/Local/Programs/Quarto/bin/quarto.js:38650:15)
    at eventLoopTick (ext:core/01_core.js:175:7)
    at async file:///C:/Users/jonat/AppData/Local/Programs/Quarto/bin/quarto.js:38612:22
    at async Semaphore.runExclusive (file:///C:/Users/jonat/AppData/Local/Programs/Quarto/bin/quarto.js:38287:29)
    at async Object.createPngsFromHtml (file:///C:/Users/jonat/AppData/Local/Programs/Quarto/bin/quarto.js:38877:42)
    at async makePng (file:///C:/Users/jonat/AppData/Local/Programs/Quarto/bin/quarto.js:81797:70)
    at async makeDefault (file:///C:/Users/jonat/AppData/Local/Programs/Quarto/bin/quarto.js:81857:24)
    at async Object.cell (file:///C:/Users/jonat/AppData/Local/Programs/Quarto/bin/quarto.js:81880:20)
    at async Promise.all (index 0)
    at async Object.document (file:///C:/Users/jonat/AppData/Local/Programs/Quarto/bin/quarto.js:39172:25)
```

In this issue above, I said it **_hangs_**, but just noticed it technically does not hang, but **blocks for just shy of 3 minutes** and ends with the above stack trace. After this I just noticed that **a PDF is generated** and contains the expected diagram! It just takes a few minutes. The test diagram and qmd document is tiny (show above), so it's not a data processing issue.

Two weeks ago attempted at trying to force a recent Deno version, but was was wisly advised here to walk away from that edge. I was hoping recent builds of Quarto using upgraded Deno 1.46.0+ would solve the problem. I have tried a few other things like reinstalling Chrome.

I'd be happy to help diagnose, repro, and try suggestions. 💛 Should I/we should open a bug at deno?

**Comment by cderv** [comment]
> Same error with: export DENO_TLS_CA_STORE=mozilla

@javajon I did not offer to try it, because currently my understanding is that Quarto won't respect a value set by the user in the environment. Quarto sets its own value, which overwrite the one in the outside environment. 

This is why I am thinking we could offer a way to do what you just did. 

>  ends with the above stack trace

This stack trace is related to chrome and Quarto usage or remote interface. Maybe the the Certificate is no more a problem for Deno, and now the "hang" issue is about chrome. 

Chrome recently change their headless browser and it as caused issues especially on windows. 

> After this I just noticed that a PDF is generated and contains the expected diagram!

Though this is very strange that you get a stack trace from chrome process error and still get a PDF. 

> I was hoping recent builds of Quarto using upgraded Deno 1.46.0+ would solve the problem.

I think it solved the Deno TLS issue panic with certificate. 

>  I have tried a few other things like reinstalling Chrome.

Which version of chrome are you on ? Quarto is supposed to use 
* `chrome.exe` on Windows found through registry (`HKLM\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\App Paths\\`
* value from `HKCR\\ChromeHTML\\shell\\open\\command`
* or the version downloaded at `quarto install chromium`

> Should I/we should open a bug at deno?

Let me try update my chrome version. I more inclined on a chrome problem that Deno problem now.

**Comment by cderv** [comment]
OK so now with latest chrome on windows I can reproduce what you see. 
````
==> quarto preview index.qmd --to typst --no-watch-inputs --no-browse

ERROR: Couldn't find open server.
Chrome process error: 

Stack trace:
Chrome process error: 
    at criClient (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/core/cri/cri.ts:100:11)
    at eventLoopTick (ext:core/01_core.js:175:7)
    at async file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/core/cri/cri.ts:58:16
    at async Semaphore.runExclusive (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/core/lib/semaphore.ts:42:22)
    at async Object.createPngsFromHtml (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/core/handlers/base.ts:180:36)
    at async makePng (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/core/handlers/mermaid.ts:341:11)
    at async makeDefault (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/core/handlers/mermaid.ts:435:16)
    at async Object.cell (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/core/handlers/mermaid.ts:475:14)
    at async Promise.all (index 0)
    at async Object.document (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/core/handlers/base.ts:577:20)
````

This is with `130.0.6723.59` and previous I had version 129 (don't remember the other numbers). 

It confirms this is not Deno, and not related to the CA store. For this one, update to deno 1.46.3 in Quarto solves it, in a way that deno now throw a warning an proceed, which panicking before that. 

Thanks a lot for helping us find this issue with new chrome !

**Comment by cderv** [comment]
Trying to debug, I can see now that locally this errors 
https://github.com/quarto-dev/quarto-cli/blob/8e429f00c5ad923602deff3894fb8a9d32b224e3/src/core/cri/cri.ts#L89

But we don't check for status, and we don't fail. I don't see error message... I'll keep debugging. I think we need to call chrome differently, especially now that they mixes the headless feature inside the main browser.

**Comment by cderv** [comment]
@javajon do you have a chrome window opened when doing `quarto render` ? Can you try with being sure all chrome windows are closed ? 

I believe the new headless mode (https://developer.chrome.com/docs/chromium/headless) will use current opened session if one. They just recently fixed issue with windows too (https://issues.chromium.org/issues/362545030) 

So it may still be issue, but I think we can call chrome with new mode correctly. Otherwise, we may rely on `--headless=old` to get back previous behavior. But from v132 it will not exist anymore

**Comment by lei0zhou** [comment]
macos 15.0.1 also has this [issue](https://github.com/quarto-dev/quarto-cli/issues/10573#issuecomment-2429991986)

**Comment by cderv** [comment]
Probably a change in chrome 130 yes then. Thanks for confirming it is not only Windows.

**Comment by javajon** [comment]
@cderv Your clarifications and confirmations are very helpful.

So this issue covers two unrelated problems:

1. "Unable to add system certificate to certificate store: InvalidCertificate"
2. Quarto to Chrome access timeout due to alleged Chrome defect with recent updates.

> Though this is very strange that you get a stack trace from chrome process error and still get a PDF.

For the small test, the PDF was generated after the long delay and stack trace. However, when processing a larger qmd with Mermaid it failed the same way and did not generate a PDF.

> Which version of chrome are you on? 

My Chrome version on Windows 11 was 
Version 130.0.6723.59 (Official Build) (64-bit) 
it just automatically updated to 
Version 130.0.6723.70 (Official Build) (64-bit)

> do you have a chrome window opened when doing quarto render? Can you try with being sure all chrome windows are closed?

Indeed I was leaving Chrome open (same one in the registry path) when generating these PDFs. It's good to know there is an interwoven dependency during this process.

You are correct, no error or significant delay was present when rendering without Chrome running. A valid PDF with the mermaid diagram was generated. Serenity now.

```bash
quarto preview c:/dev/trinity-courses/test/test-graph-to-pdf.qmd --to pdf --no-browser --no-watch-inputs
Unable to add system certificate to certificate store: InvalidCertificate(Other(OtherError(UnsupportedCriticalExtension)))
308203b33082029ba00302010[...]384095e3
pandoc 
  to: latex
  output-file: test-graph-to-pdf.tex
  standalone: true
  pdf-engine: xelatex
  variables:
    graphics: true
    tables: true
  default-image-extension: pdf

metadata
  documentclass: scrartcl
  classoption:
    - DIV=11
    - numbers=noendperiod
  papersize: letter
  header-includes:
    - \KOMAoption{captions}{tableheading}
  block-headings: true

Rendering PDF
running xelatex - 1
  This is XeTeX, Version 3.141592653-2.6-0.999996 (TeX Live 2024) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

running xelatex - 2
  This is XeTeX, Version 3.141592653-2.6-0.999996 (TeX Live 2024) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

Watching files for changes
Watching files for changes
Browse at http://localhost:3922/web/viewer.html
```

Without Chrome running the above process created a healthy test-graph-to-pdf.pdf.

I started Chrome and the error with a long pause and the stack trace came back.

I exited Chrome and the PDF rendered again without an error. 

With Chrome closed, I tried to render my larger qmd with extensions, multiple slides, and three mermaids to PDF. This process just hung. I waited over 10 minutes, still stuck. The RevealJS render works fine.  Moreover, while stuck on this larger PDF render or even after Ctrl-C I could not start Chrome on my desktop. A reboot cleared the Chrome startup problem. So not having Chrome running sort of solved the problem. (Or, perhaps the larger qmd is exposing a third problem.)

For now, I will render PDF without Chrome running and await or go hunt for the dev v132 Chrome release. Perhaps running Firefox is another way to avoid these rendering issues with Chrome.

**Comment by javajon** [comment]
> But from v132 it will not exist anymore

Confirmed with Chrome Version 132.0.6779.0 (Official Build) dev (64-bit) running on desktop only, then the render works. Closed that Chrome and ran Chrome Version 130.0.6723.70 (Official Build) (64-bit) and the rendering produced a stack trace error.

**Comment by javajon** [comment]
With Chrome Version 132.0.6779.0 the PDF rendering the above qmd with 1 graph works.

The same Chrome and render to PDF hangs with this qmd with three Mermaids.  

```markdown

## Graph One

```{mermaid}
%%| echo: false
%%| fig-width: 5.5
%%{init: {'theme':'dark'}}%%
graph TB
    A((8))-->B((3))
    A-->C((10))
    B-->D((1))
    B-->E((6))
    C-->F((9))
    C-->G((14))
    E-->H((4))
    E-->I((7))
```

## Graph Two

```{mermaid}
%%| echo: false
%%| fig-width: 5.5
%%{init: {'theme':'dark'}}%%
graph TB
    A((8))-->B((3))
    A-->C((10))
    B-->D((1))
    B-->E((6))
    C-->F((9))
    C-->G((14))
    E-->H((4))
    E-->I((7))
```

## Graph Three

```{mermaid}
%%| echo: false
%%| fig-width: 5.5
%%{init: {'theme':'dark'}}%%
graph TD
    A('*')
    B('+')
    C('-')
    D[5]
    E[2]
    F[8]
    G[3]
    A --> B
    A --> C
    B --> D
    B --> E
    C --> F
    C --> G
```

Using same ommand `quarto preview test-three-graphs-to-pdf.qmd --to pdf`. Perhaps the v132 does not fully address he problem or there is somethign else going on. @cderv Can you see if this three Mermaid qmd to PDF hangs for you too?

**Comment by cderv** [comment]
I marked this issue for discussion with team. We probably need to adapt the way we call chrome to newest logic from them. 

What I meant by v132 changing, is that Chromium team informed that the `--headless=old` support will be removed as they are working on making an extenal binary for the old headless tool.

I don't have v132, but I noticed that v130 works of with Quarto 1.5 release. So this is probably also to fixed we tried initial when v128 made the change to `--headless=new` as a default for `--headless` and it is not working any more. On chromium tracker some issues are still open, so v130 may have some fixes, but v132 others and not fully working. 

We'll be coming with an update on this, and report back. Thanks for your patience !
